### PR TITLE
Performance: reduce memory and I/O overhead of the document exporter during zip exports

### PR DESCRIPTION
**Pull request by stumpylog**
## Proposed change

The export to a zip file was always kind of grafted on.  The main problems were the double file IO: we wrote to the temorary directory, then zipped it, then copied it to the output.  Kind of fragile and doing about 2x the IO it really needed

This uses the general idea from the bulk download strategy, with an `ExportSink` abstract class and concrete implementations for a directory sink and a zip sink.  The class supports adding a simple file, a JSON file by content and name, and finally a streaming file, for the overall manifest.  In the future, it could be extended to other ideas, like FTP maybe or S3.

This also sets up for a future branch I have, to allow controlling aspects of the zip compression.  And it provides a base to I think eventually allow a web based export via the zip sink.

Except for patching changes all existing exporter tests work the same.  (`--zip` combined with `--compare-*` flags is now explicitly rejected, since that combination never made sense.)

It includes a little fix/improvement to use the QuerySetStream class to stream Documents into the main dict, which reduced memory usage to about 1/12th of what it was before.

## Type of change

- [ ] Bug fix: non-breaking change which fixes an issue.
- [x] New feature / Enhancement: non-breaking change which adds functionality. _Please read the important note above._
- [ ] Breaking change: fix or feature that would cause existing functionality to not work as expected.
- [ ] Documentation only.
- [ ] Other. Please explain:

## Checklist:

- [ ] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [ ] If applicable, I have included testing coverage for new code in this PR, for [backend](https://docs.paperless-ngx.com/development/#testing) and / or [front-end](https://docs.paperless-ngx.com/development/#testing-and-code-style) changes.
- [ ] If applicable, I have tested my code for breaking changes & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [ ] If applicable, I have checked that all tests pass, see [documentation](https://docs.paperless-ngx.com/development/#back-end-development).
- [ ] I have run all Git `pre-commit` hooks, see [documentation](https://docs.paperless-ngx.com/development/#code-formatting-with-pre-commit-hooks).
- [ ] I have made corresponding changes to the documentation as needed.
- [ ] In the description of the PR above I have disclosed the use of AI tools in the coding of this PR.

### CODE REVIEW

File: src/documents/export/sinks.py [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+            ):
+                tmp.unlink()
+                return
+        tmp.rename(target)
+
+    def _finalize(self) -> None:
+        if self._delete:
+            for f in self._snapshot:
+                f.unlink()
+                delete_empty_directories(f.parent, self._target)
```
[the remark this answers is not in the document]
Comment (stumpylog) [reply src/documents/export/sinks.py:230]: Technically correct I suppose, but it also feels like an edge case.  The code the same as before though.

**Review by shamoon — APPROVED** [review]
Will go with the copilot review and lots of working tests here
